src/rust/bitbox02-rust/src/main_loop.rs

```diff
@@ -0,0 +1,190 @@
+                // unlock it.
+                if u2f_data.is_none() && bitbox02::usb_processing::locked_u2f() {
+                    bitbox02::usb_processing::unlock();
+                }
+            }
+        }
+
+        // Do USB Input
+        if hww_data.is_none() && bitbox02::hid_hww::read(&mut hww_frame) {
+            if bitbox02::usb_packet::process(&hww_frame) {
+                if bitbox02::communication_mode::ble_enabled() {
+                    // Enqueue a power down command to the da14531
+                    bitbox02::da14531::power_down(&mut uart_write_queue);
+                    // Flush out the power down command. This will be the last UART communication
+                    // we do.
+                    while uart_write_queue.len() > 0 {
+                        bitbox02::uart::poll(None, None, Some(&mut uart_write_queue));
+                    }
+                    bitbox02::communication_mode::ble_disable();
+                }
+            } else {
+                log!("usb_packet_process: invalid");
+            }
+        }
+        #[cfg(feature = "app-u2f")]
+        if u2f_data.is_none() && bitbox02::hid_u2f::read(&mut u2f_frame) {
+            bitbox02::u2f_packet::process(&u2f_frame);
+        }
+
+        // Do UART Output
+        if bitbox02::communication_mode::ble_enabled() {
+            if let Some(frame) = bitbox02::da14531_protocol::poll(
+                &mut uart_read_buf,
+                &mut uart_read_buf_len,
+                &mut hww_data,
+                &mut uart_write_queue,
+            ) {
+                log!("got frame, calling handler");
+                bitbox02::da14531_handler::handler(frame, &mut uart_write_queue);
+            }
+        }
+
+        // Do USB Output
+        if let Some(data) = &mut hww_data
+            && !bitbox02::communication_mode::ble_enabled()
+        {
+            if bitbox02::hid_hww::write_poll(data) {
+                hww_data = None;
+            }
+        }
+        #[cfg(feature = "app-u2f")]
+        if let Some(data) = &mut u2f_data {
+            if bitbox02::hid_u2f::write_poll(data) {
+                u2f_data = None;
+            }
+        }
+
+        /* First, process all the incoming USB traffic. */
+        bitbox02::usb_processing::process_hww();
+        #[cfg(feature = "app-u2f")]
+        bitbox02::usb_processing::process_u2f();
+
+        /*
+         * If USB has generated events at the application level,
+         * process them now.
+         */
+        #[cfg(feature = "app-u2f")]
+        bitbox02::u2f::process();
+
+        bitbox02::screen::process();
+
+        /* And finally, run the high-level event processing. */
+        #[cfg(feature = "app-u2f")]
+        unsafe {
+            crate::workflow::u2f_c_api::rust_workflow_spin()
+        }
+        crate::async_usb::spin();
+
+        if let Some(ref mut task) = orientation_task {
+            if let Poll::Ready(_orientation) = util::bb02_async::spin(task) {
```

Comment:
Just fyi, you can now add more conditions, like `if let ... && let ...`. Above you also have some double nestings becaue of that, which you could flatten.